drivers/ethernet/eth_w5500.c

```diff
@@ -342,28 +342,49 @@ static int w5500_set_config(const struct device *dev,
 {
 	uint8_t mode;
 	uint8_t mr = W5500_S0_MR_MF;
+	struct w5500_runtime *ctx = dev->data;
 
-	w5500_spi_read(dev, W5500_S0_MR, &mode, 1);
-	if (IS_ENABLED(CONFIG_NET_PROMISCUOUS_MODE) &&
-	    type == ETHERNET_CONFIG_TYPE_PROMISC_MODE) {
-		if (config->promisc_mode) {
-			if (!(mode & BIT(mr))) {
+	switch (type) {
+	case ETHERNET_CONFIG_TYPE_MAC_ADDRESS:
+		memcpy(ctx->mac_addr,
+			config->mac_address.addr,
+			sizeof(ctx->mac_addr));
+		w5500_spi_write(dev, W5500_SHAR, ctx->mac_addr, sizeof(ctx->mac_addr));
+		LOG_INF("%s MAC set to %02x:%02x:%02x:%02x:%02x:%02x",
+			dev->name,
+			ctx->mac_addr[0], ctx->mac_addr[1],
+			ctx->mac_addr[2], ctx->mac_addr[3],
+			ctx->mac_addr[4], ctx->mac_addr[5]);
+
+		/* Register Ethernet MAC Address with the upper layer */
+		net_if_set_link_addr(ctx->iface, ctx->mac_addr,
+			sizeof(ctx->mac_addr),
+			NET_LINK_ETHERNET);
+
+		return 0;
+	default:
+		w5500_spi_read(dev, W5500_S0_MR, &mode, 1);
+		if (IS_ENABLED(CONFIG_NET_PROMISCUOUS_MODE) &&
+		    type == ETHERNET_CONFIG_TYPE_PROMISC_MODE) {
+			if (config->promisc_mode) {
+				if (!(mode & BIT(mr))) {
+					return -EALREADY;
+				}
+			}
+
+			/* clear */
+			WRITE_BIT(mode, mr, 0);
+		} else {
+			if (mode & BIT(mr)) {
 				return -EALREADY;
 			}
-		}
 
-		/* clear */
-		WRITE_BIT(mode, mr, 0);
-	} else {
-		if (mode & BIT(mr)) {
-			return -EALREADY;
+			/* set */
+			WRITE_BIT(mode, mr, 1);
 		}
 
-		/* set */
-		WRITE_BIT(mode, mr, 1);
+		return w5500_spi_write(dev, W5500_S0_MR, &mode, 1);
```

Comment:
@jfischer-no took your changes and added some comments. 

Also changed the w5500_hw_start() so that W5500 is started by default with promiscuous mode disabled (hw MAC filtering enabled). So far hw promiscuous mode was active from start even if CONFIG_NET_PROMISCUOUS_MODE was disabled. Don't know if this could pose any security risks in the past.

Don't have any tools to test promiscuous mode separately and this part of the code wasn't tested.

Reply:
Moved promiscuous mode changes to separate commit.